### PR TITLE
DPL-1077 - PacBio run volume checks

### LINKED ISSUE #1518: DPL-1077 - As a long read user I want to see an automatic volume check when I add a library to a sequencing run to ensure the integrity of the data is maintained
**User story**
As a long read user I want to see an automatic volume check when I add a library to a sequencing run to ensure the integrity of the data is maintained

**Who are the primary contacts for this story**
Ben / Steve / James / Seena

**Who is the nominated tester for UAT**
James / Long Read

**Acceptance criteria**
To be considered successful the solution must allow:
- [ ] the ability to add the volume required when I add a library to a well in a sequencing run
- [ ] an automated check to ensure that there is enough library available
- [ ] an appropriate message if there is not enough
- [ ] a success message if the volume is accepted and the library is added

RunCreate store structure
- Currently in PacBio Run Create store we send a list of pool_ids and library_ids in each well that then creates the used_aliquots in the service. We should update this so we send a list of used_aliquots instead (with their new volumes) and change the API to remove pool_ids and library_ids. See https://github.com/sanger/traction-service/pull/1274#discussion_r1559041462 for more information

**Dependencies**
This story is blocked by the following dependencies:
- sanger/traction-service#1186
- sanger/traction-service#1136

References
This story has a non-blocking relationship with:
- #<issue_no.>
- sanger/<repo>#<issue_no.>

**Additional context**
- Volume checking process map: [https://lucid.app/lucidchart/3c12fbb3-5265-49b5-9f6a-97c9a69a59eb/edit?viewport_l[…]3271%2C0_0&invitationId=inv_4a6b1f8d-e886-4ffc-b784-40f2135c0679](https://lucid.app/lucidchart/3c12fbb3-5265-49b5-9f6a-97c9a69a59eb/edit?viewport_loc=116%2C-4472%2C6583%2C3271%2C0_0&invitationId=inv_4a6b1f8d-e886-4ffc-b784-40f2135c0679)
- Multiplexing ERD: https://lucid.app/lucidchart/56eec268-2911-44c4-ab4b-87393e811738/edit?view_items=Fm2l5AbOONXb&invitationId=inv_81ba033e-8594-44c3-a37f-fd0c07fb4803
- Multiplexing process map: https://lucid.app/lucidchart/343de76c-d3e5-4a26-92c4-7bc10d919fb8/edit?page=0_0&invitationId=inv_13a8dd3b-ef1b-4f4e-b160-0dbf45a6031f#

### PR DESCRIPTION
**Pull request by BenTopping**
Closes #1518, #1617

#### Changes proposed in this pull request

- Updates PacBio Run Well and associated components to handle used_aliquots instead of pool_ids and library_ids to better setup volume handling.
- Fixes run well colour for SMRTLink v13 wells.
- Adds volume input field to pools and libraries in PacBio run wells.
- Adds available_volume badge to wells for aliquots sourced from libraries.

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_  
_[Production Merges to `main`]_  
 &nbsp; &nbsp; \- _Check story numbers included_  
 &nbsp; &nbsp; \- _Check for debug code_  
 &nbsp; &nbsp; \- _Check version_

### CODE REVIEW

File: src/stores/pacbioRunCreate.js [hunk 1]
```diff
@@ -508,5 +515,73 @@ export const usePacbioRunCreateStore = defineStore('pacbioRunCreate', {
...
+      let total_available_volume = (
+        parseFloat(library_available_volume) - parseFloat(library_used_aliquots_volume)
+      ).toFixed(2)
+
+      // If its an existing aliquot we need to add the original volume back
+      // Because its taken into account in the library_available_volume
+      // Unless the volume is 0 as that won't affect the available volume
+      if (original_aliquot && original_aliquot.volume != 0) {
+        total_available_volume = (
+          parseFloat(total_available_volume) + parseFloat(original_aliquot.volume)
```
Comment (seenanair) [inline src/stores/pacbioRunCreate.js:574]: I believe we don't need to use parseFloat as they are already numbers.

File: src/stores/pacbioRunCreate.js [hunk 2]
```diff
@@ -508,5 +515,73 @@ export const usePacbioRunCreateStore = defineStore('pacbioRunCreate', {
...
+      const original_aliquot = this.aliquots[aliquotId]
+      // Get the available volume for the library
+      const library_available_volume = this.libraries[libraryId].available_volume || 0
+
+      // Calculate the sum of the volume of all the new aliquots used in wells that are from the library
+      let library_used_aliquots_volume = 0
+      // eslint-disable-next-line no-unused-vars
+      Object.entries(this.wells).forEach(([_plateNumber, plate]) => {
+        // eslint-disable-next-line no-unused-vars
+        Object.entries(plate).forEach(([_position, well]) => {
```
Comment (seenanair) [inline src/stores/pacbioRunCreate.js:541]: Would it be possible to  use Object.values instead as we are only interested those and also avoids eslint complains as well?

File: src/stores/pacbioRunCreate.js [hunk 3]
```diff
@@ -508,5 +515,73 @@ export const usePacbioRunCreateStore = defineStore('pacbioRunCreate', {
...
+
+      // Get the original aliquot if it exists
+      const original_aliquot = this.aliquots[aliquotId]
+      // Get the available volume for the library
+      const library_available_volume = this.libraries[libraryId].available_volume || 0
+
+      // Calculate the sum of the volume of all the new aliquots used in wells that are from the library
+      let library_used_aliquots_volume = 0
+      // eslint-disable-next-line no-unused-vars
+      Object.entries(this.wells).forEach(([_plateNumber, plate]) => {
```
Comment (seenanair) [inline src/stores/pacbioRunCreate.js:539]: Just a thought - may be reduce  instead of forEach in such situations, but then it's not a big deal

File: src/stores/pacbioRunCreate.js [hunk 4]
```diff
@@ -508,5 +515,73 @@ export const usePacbioRunCreateStore = defineStore('pacbioRunCreate', {
...
     },
+    /**
+     * Returns the available volume for a library aliquot
+     * @param {Object} libraryId The id of the library
+     * @param {Object} aliquotId The id of the aliquot
+     * @param {Object} volume The volume of the aliquot
+     *
+     * @returns {Number} The available volume for the aliquot
+     */
+    getAvailableVolumeForLibraryAliquot({ libraryId = null, aliquotId = null, volume = null }) {
```
Comment (seenanair) [inline src/stores/pacbioRunCreate.js:526]: Good job, it looks like you have covered it all 💯

**Review by seenanair — COMMENTED** [review]
Great job, moving to aliquots instead of pools and libraries, reduced the complexity a lot 👏 
Just added a few thoughts, nothing major.

File: src/components/labware/PacbioRunWell.vue [hunk 5]
```diff
@@ -122,20 +113,25 @@ const storeWell = computed(() => {
...
-  return [...storeWell.value.pools, ...storeWell.value.libraries]
-    .map((id) => {
-      return store.tubeContents.find((tubeContent) => id == tubeContent.id).barcode
-    })
-    .filter(Boolean)
-    .join(',')
+  return storeWell.value?.used_aliquots
+    ? [...storeWell.value.used_aliquots]
+        .map(({ source_id, source_type }) => {
+          const type = source_type === 'Pacbio::Pool' ? 'pools' : 'libraries'
```
[the remark this answers is not in the document]
Comment (BenTopping) [reply src/components/labware/PacbioRunWell.vue:119]: Ill have a look into it it but im not sure about forcefully changing the type because we rely on it in a few places like in the badge of the items the pool/library list and when linking data together.

File: src/stores/pacbioRunCreate.js (hunk 3 above)
Comment (BenTopping) [reply src/stores/pacbioRunCreate.js]: Yep reduce works too, I tried it originally but I was finding it hard to follow with the nesting.

File: src/stores/pacbioRunCreate.js [hunk 6]
```diff
@@ -508,5 +515,73 @@ export const usePacbioRunCreateStore = defineStore('pacbioRunCreate', {
...
+        ).toFixed(2)
+      } else {
+        // If its a new aliquot we need to add the volume back in to the total available volume
+        // because it was removed as part of the library_used_aliquots volume but it should be available in this instance of aliquot
+        total_available_volume = (parseFloat(total_available_volume) + parseFloat(volume)).toFixed(
+          2,
+        )
+      }
+
+      return total_available_volume
```
[the remark this answers is not in the document]
Comment (BenTopping) [reply src/stores/pacbioRunCreate.js]: Yep that makes it cleaner 👌

File: src/stores/pacbioRunCreate.js (hunk 3 above)
Comment (seenanair) [reply src/stores/pacbioRunCreate.js]: Yes, true. I also find that sometimes `reduce` 'reduces' readability :)

File: src/components/labware/PacbioRunWell.vue (hunk 5 above)
[the remark this answers is not in the document]
Comment (BenTopping) [reply src/components/labware/PacbioRunWell.vue:119]: Had a look - we are using that type to find where to look in the store in a few places so it we made this change we would essentially have to do the reverse in some other places to fix it so think its best to leave.

[the remark this answers is not in the document]
Comment (seenanair) [reply src/components/labware/PacbioRunWell.vue:119]: Cool, That make sense. No worries 👍

**Comment by seenanair** [comment]
@BenTopping While checking, I found this issue: if we drop a library or pool into a well and then remove it, it is still displayed in red. Is this intended? I thought that since it is empty, it should revert to its original white colour.

**Comment by BenTopping** [comment]
@seenanair Yes thats intended. When you drop a library/pool it creates the well and populates some of the metadata with defaults alongside adding the library/pool. Since a user may remove a library/pool and then add another one and keep the same metadata they have entered it makes sense to keep the well. The well is only deleted when the 'delete well' button is pressed.

**Review by seenanair — APPROVED** [review]
